### PR TITLE
add strong typing to root.stringtable by making it a template

### PR DESCRIPTION
**Pull request by benjones**
Gets rid of a lot of casts.  Also allows storing slices as values rather than just pointers.

### CODE REVIEW

File: src/dmd/filecache.d [hunk 1]
```diff
@@ -108,11 +108,11 @@ struct FileCache
         if (auto payload = files.lookup(file))
         {
             if (payload !is null)
-                return cast(typeof(return)) payload.ptrvalue;
+                return  payload.value;
```
Comment (rainers) [inline src/dmd/filecache.d]: There is an extra space here.

File: src/dmd/libelf.d [hunk 2]
```diff
@@ -364,7 +364,7 @@ private:
                 scanObjModule(om);
             }
         }
-        /************* Determine string section ******************/
+        /************* Determine  section ******************/
```
Comment (rainers) [inline src/dmd/libelf.d]: Why remove part of the comment here?

File: src/dmd/traits.d [hunk 3]
```diff
@@ -86,7 +86,7 @@ private Dsymbol getDsymbolWithoutExpCtx(RootObject oarg)
     return getDsymbol(oarg);
 }
 
-private const StringTable traitsStringTable;
+private const StringTable!string traitsStringTable;
```
Comment (rainers) [inline src/dmd/traits.d]: No need to duplicate the key by the value, you can just use StringTable!bool and print `sub.lstring()` below.

**Review by rainers — COMMENTED** [review]
Apart from the nits, LGTM.

File: src/dmd/libmscoff.d [hunk 4]
```diff
@@ -71,12 +71,6 @@ final class LibMSCoff : Library
 {
     MSCoffObjModules objmodules; // MSCoffObjModule[]
     MSCoffObjSymbols objsymbols; // MSCoffObjSymbol[]
-    StringTable tab;
```
Comment (Geod24) [inline src/dmd/libmscoff.d]: Nicely spotted

**Review by Geod24 — APPROVED** [review]

**Comment by Geod24** [comment]
Sweet

**Pull request merged**